python/tank/bootstrap/configuration.py

```diff
@@ -154,51 +175,80 @@ def _set_authenticated_user(self, user, serialized_user):
             from ..util import CoreDefaultsManager
         except ImportError:
             log.debug("Using pre-0.16 core, no authenticated user will be set.")
-            return
+            return None
+
+        from ..authentication import deserialize_user
+        from .. import api
 
         log.debug("The project's core supports the authentication module.")
 
-        # Check to see if there is a user associated with the current project.
+        # Retrieve the user associated with the current project.
         default_user = ShotgunAuthenticator(CoreDefaultsManager()).get_default_user()
 
-        # Assume we'll use the same user as was used for bootstrapping to authenticate.
-        authenticated_user = user
-        # If we have a default user...
+        # By default, we'll assume we couldn't find a user with the authenticator.
+        project_user = None
+
+        # If the project core's authentication code found a user...
+        # (Note that in the following code, a user with no login is a script user.)
         if default_user:
-            # ... and it doesn't have a login
+            # If the project uses a script user, we'll use that.
             if not default_user.login:
-                log.debug("Script user found for this project.")
-                # it means we're dealing with a script user and we'll use that, so override
-                # the authenticated user.
-                authenticated_user = default_user
-            else:
-                # We found a user, but we'll ignore it.
+                log.debug("User retrieved for the project is a script user.")
+                project_user = default_user
+            # If the project didn't use a script, but the bootstrap did, we'll keep using it.
+            elif not bootstrap_user_login:
+                # We'll keep using the bootstrap user. This is because configurations like tk-
+                # config-basic or tk-config-default are meant to be used with whatever credentials
+                # were used during bootstrapping when used with a CachedDescriptor. The bootstrap
+                # user is a script user and the project's user is not a script user, so we'll keep
+                # using the script user.
+
+                # Note that if the API endpoint is different in the project it means that we'll
```

Comment:
what do you mean when you say API endpoint here?

Reply:
I'm talking about hosted sites who have multiple endpoints, one to query the webapp, one or more to query with the API.

Reply:
i think we can still clarify this further - the webapp and API always use different endpoints for traffic. Maybe we can just say 'handle the case when a local install has got separate dedicated webapp servers for webui and API traffic". Or something along those lines. I think that's much clearer.